core/Network/TLS/Handshake/Server.hs

```diff
@@ -315,9 +315,24 @@ doHandshake sparams mcred ctx chosenVersion usedCipher usedCompression clientSes
             usingHState ctx $ setDHPrivate priv
             return serverParams
 
+        decideHash sigAlg = do
+            usedVersion <- usingState_ ctx getVersion
+            case usedVersion of
+              TLS12 -> do
+                  let sHashSigs = supportedHashSignatures $ ctxSupported ctx
+                      cHashSigs = case extensionLookup extensionID_SignatureAlgorithms exts >>= extensionDecode False of
+                          Just (SignatureAlgorithms hss) -> hss
+                          Nothing                        -> []
+                      hashSigs = cHashSigs `intersect` sHashSigs
+                  case filter ((==) sigAlg . snd) hashSigs of
+                      []  -> error ("no hash signature for " ++ show sigAlg)
+                      x:_ -> return $ Just (fst x)
```

Comment:
Does this always the algorithm that is most preferred by the client?
If the Haskell server can be configured to preempt the client's cipherlist and choose the cipher it prefers most, then it makes sense to also use the server preference order for signature algorithms.